File: 0581-Shortest Unsorted Continuous Subarray/0581-Shortest Unsorted Continuous Subarray.cpp
class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
        int n = nums.size();
        stack<int> St1;
        int first = n + 1, last = -1;
        for (int i = 0; i < n; ++i)
        {
            while (!St1.empty() && nums[St1.top()] > nums[i])
            {
                first = min(first, St1.top());
                St1.pop();
            }
            St1.push(i);
        }
        
        if (first == n + 1)
        {
            return 0;
        }
        
        stack<int> St2;
        for (int i = n - 1; i >= 0; --i)
        {
            while (!St2.empty() && nums[St2.top()] < nums[i])
            {
                last = max(last, St2.top());
                St2.pop();
            }
            St2.push(i);
        }

        return last - first + 1;
    }
};

// O(1) space
class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
        int n = nums.size();
        bool flag = false;
        int minNum = INT_MAX, maxNum = INT_MIN;
        for (int i = 1; i < n; ++i)
        {
            if (nums[i] < nums[i - 1])
            {
                flag = true;    
            }
            
            if (flag)
            {
                minNum = min(minNum, nums[i]);
            }
        }

        flag = false;        
        for (int i = n - 2; i >= 0; --i)
        {
            if (nums[i] > nums[i + 1])
            {
                flag = true;    
            }
            
            if (flag)
            {
                maxNum = max(maxNum, nums[i]);
            }
        }

        int l, r;
        for (l = 0; l < n; ++l)
        {
            if (nums[l] > minNum)
            {
                break;
            }
        }
        
        for (r = n - 1; r >= 0; --r)
        {
            if (nums[r] < maxNum)
            {
                break;
            }
        }
        
        return r - l < 0 ? 0 : r - l + 1; 
    }
};

File: 0501-0600/0581-Shortest Unsorted Continuous Subarray/0581-Shortest Unsorted Continuous Subarray.py
import math

class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        snums = sorted(nums)
        l = 0
        while l < len(nums):
            if nums[l] != snums[l]:
                break
            l += 1
        if l == len(nums):
            return 0
        r = len(nums) - 1
        while r >= 0:
            if nums[r] != snums[r]:
                break
            r -= 1
        return r - l + 1 if r >= l else 0

class Solution2:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        l = 0
        r = -1
        maxNum = nums[0]
        minNum = nums[-1]
        for i in range(1, len(nums)):
            maxNum = max(maxNum, nums[i])
            minNum = min(minNum, nums[-i-1])
            if nums[i] < maxNum:
                r = i
            if nums[-i-1] > minNum:
                l = len(nums) - i - 1
        return r - l + 1
